get_col_for_pattern <-function(this_pattern)
{
  split_pattern <- strsplit(this_pattern, "")[[1]]
  # pattern = uuuuuuuuuuuuuuuuuuuuMuuuuuuuuuuuuuuMuuuuuuuuuuuuuuuuuuuuuMMPM
  psize         <- length(split_pattern)
  pattern_col   <- matrix(, nrow = 2, ncol = psize)
  #
  for (i in 1:psize)
  {
    if(split_pattern[i] == 'u') 
    {
      pattern_col[1, i] <- '0'
      pattern_col[2, i] <- 'gray'
    }
    else
      if(split_pattern[i] == 'P') 
      {
        pattern_col[1, i] <- 'p'
        pattern_col[2, i] <- 'darkblue'
      }
    else      
      if(split_pattern[i] == 'M') 
      {
        pattern_col[1, i] <- 'm'
        pattern_col[2, i] <- 'red'
      }
    else
    {
      pattern_col[1, i] <- split_pattern[i] # case of alleles
      pattern_col[2, i] <- 'purple'         # not important
    }
  }
  return(pattern_col)
}
#
get_contig_id <- function(contiginfo)
{
  # contiginfo = s1_genotype_pollen_seq_contig_tig00000013_pilon.txt
  contig <- strsplit(contiginfo, '_')
  if(length(contig[[1]]) < 7)
    stop("Error: file name about contig sequence and pattern must be 7 bits, as contig name need to be extracted stringently. \n")
  contig2 <- strsplit(paste(contig[[1]][6], contig[[1]][7],sep="_"), '.txt')    
  return(contig2[[1]][1])
}
#!/bin/env Rscript

## get options from cmd
###############################################################################
# Rscript ../../visualize_pollen_cluster.R /biodata/dep_coupland/grp_schneeberger/projects/methods/src_shq/z_asPollinator_2019/z_test_ic120_ac280_iafp3_aafp7_total250550_0924_tmp_pollen_genotypes/s1_genotype_pollen_seq_ctgwise/s1_genotype_pollen_seq_contig_tig00000013_pilon.txt
# cd /biodata/dep_coupland/grp_schneeberger/projects/methods/src_shq/z_asPollinator_2019/z_test_ic120_ac280_iafp3_aafp7_total250550_0924_tmp_pollen_genotypes/s1_genotype_pollen_seq_ctgwise/
# for i in *.txt; do Rscript ../../visualize_pollen_cluster.R ${i}; done
args<-commandArgs(TRUE)
if(length(args) < 3) {
  cat(length(args))
  cat("\n   Info: visualize clustered pollen genotypes from a file generated by asPollinator.")
  cat("   USAGE: visualize_pollen_cluster clusters markerSet contigSizes\n\n")
}else
{
  ## check args
  clusterfile = args[[1]] # "/biodata/dep_coupland/grp_schneeberger/projects/methods/src_shq/z_asPollinator_2019/test_example_for_R4_corrected.txt"
  markerfile  = args[[2]] # parental 565582: "/netscratch/dep_mercier/grp_schneeberger/projects/mutation_tree/Apricot/zzHequan_assembly_pipeline/Step_3_build_benchmark_2/marker_creation/new_marker/rojoPasion_20190819_converted_variant_ic120_ac280_iaf0.3_aaf0.7_total250550.txt"
  # phased   647881: "/netscratch/dep_mercier/grp_schneeberger/projects/mutation_tree/Apricot/zCodeTest_asPollinator_HQ/marker/20190819_converted_variant_ic120_ac280_iaf0.3_aaf0.7_total250550.txt"
  sizefile    = args[[3]] # /netscratch/dep_mercier/grp_schneeberger/projects/mutation_tree/Apricot/Pollen-b/z_related_species_reference_based/z_2019_our_Apricot_manually_curated_primary_contigs/manual_curation_our_ApricotRef/manually_curated.chrsizes_separated
  
  # prefix      = args[[3]]
  if(file.exists(clusterfile)!=TRUE)
  {
    stop(paste("Cannot find histo file: ", clusterfile, ". Please check it! Program exited.\n",sep=""))
  }
  if(file.exists(markerfile)!=TRUE)
  {
    stop(paste("Cannot find marker file: ", markerfile, ". Please check it! Program exited.\n",sep=""))
  }    
  if(file.exists(sizefile)!=TRUE)
  {
    stop(paste("Cannot find size file: ", sizefile, ". Please check it! Program exited.\n",sep=""))
  }   
  if(TRUE)
  {  
    ## read markers
    markers      <- read.table(markerfile)
    contigid     <- get_contig_id(basename(clusterfile)) # => s1_genotype_pollen_seq_contig_tig00000013_pilon.txt <= must be in this format
    # read size of contigs
    contigsize   <- read.table(sizefile)
    this_size    <- contigsize[contigsize$V1==contigid, 2]
    #
    contmarker   <- markers[markers$V2==contigid, ]      # markers for current contig
    contmarkernum<- length(contmarker$V1)
    ## read sequence/pattern
    clusters     <- read.table(clusterfile)
    m_mkr        <- clusters[clusters$V2=='0_m_mkr', ]
    p_mkr        <- clusters[clusters$V2=='0_p_mkr', ]
    this_pattern <- as.vector(m_mkr$V1)
    len_pattern  <- nchar(this_pattern)
    # check 
    if(len_pattern != contmarkernum)
      stop(paste("Error: marker number is not equal to pattern/sequence length along current contig ", contigid, "\n", sep=""))
    #
    ############################################## maternal cluster pollens
    mcluster2 <- clusters[clusters$V3=='0_m', ]
    msize     <- length(mcluster2$V1)
    if(msize > 0)
    {
      mcluster_seq     <- mcluster2[seq(1, msize, 2), ]
      mcluster_pattern <- mcluster2[seq(2, msize, 2), ]  
    }else
    {
      mcluster_seq     <- 0
      mcluster_pattern <- 0
    }
    ############################################## paternal cluster pollens
    pcluster2 <- clusters[clusters$V3=='0_p', ]
    psize     <- length(pcluster2$V1)
    if(psize > 0)
    {
      pcluster_seq     <- pcluster2[seq(1, psize, 2), ]
      pcluster_pattern <- pcluster2[seq(2, psize, 2), ]  
    }else
    {
      pcluster_seq     <- 0
      pcluster_pattern <- 0
    }
    ############################################## un-clustered pollens
    ucluster2 <- clusters[clusters$V3=='0_u', ]
    usize     <- length(ucluster2$V1)
    if(usize > 0)
    {
      ucluster_seq     <- ucluster2[seq(1, usize, 2), ]
      ucluster_pattern <- ucluster2[seq(2, usize, 2), ]  
    }else
    {
      ucluster_seq     <- 0
      ucluster_pattern <- 0
    }
    ######################################################################################################################
    ################################################# plot pdf ###########################################################
    ######################################################################################################################
    # clusterfile="/biodata/dep_coupland/grp_schneeberger/projects/methods/src_shq/z_asPollinator_2019/z_test_ic120_ac280_iafp3_aafp7_total250550_0924_tmp_pollen_genotypes/s1_genotype_pollen_seq_ctgwise/s1_genotype_pollen_seq_contig_tig00000013_pilon.txt"
    #
    totalpollennum = msize+psize+usize
    labelsizecex   = 0.6
    if(totalpollennum/2>200) labelsizecex = 0.2
    #
    pdfinfo <- tools::file_path_sans_ext(clusterfile)
    pdf(paste(pdfinfo, ".pdf", sep=""), family="Helvetica", height=12, width=8.26772)
    par(mfrow=c(3,1), mai = c(0.7, 0.5, 0.5, 0.3)); # margin: bottom, left, top, right
    #
    ############################################## visualize pollen patterns in maternal cluster
    firstmarkerpos  = min(contmarker$V3)
    lastmarkerpos   = max(contmarker$V3)
    if(length(this_size)==0) 
    {
      xlimsta = 1
      xlimend = lastmarkerpos
    }else
    {
      xlimsta = 1
      xlimend = this_size
    }
    plot(1, 1, xlim = c(xlimsta, xlimend+round(xlimend*0.2) ), ylim = c(0, totalpollennum), frame.plot = F, axes = F, xlab = "", ylab = "", col="white")
    segments(x0 = contmarker$V3, y0 = 0, x1 = contmarker$V3, y1 = totalpollennum, col="gray", lwd=0.3)
    title(main="Pollens classified in maternal cluster", cex.main = 1, col.main="black")  
    if(is.null(mcluster_pattern) == FALSE & mcluster_pattern!=0)
    {
      npollen = length(mcluster_pattern$V1)
      if(npollen > 0)
      {
        # first red: maternal truth
        for(ipat in 1:npollen)
        {
          this_seq     = strsplit( tolower(as.vector(mcluster_seq[ipat, 1]) ), "")[[1]]
          this_pattern = as.vector(mcluster_pattern[ipat, 1])
          tlen         = nchar(this_pattern)
          this_col_pat = get_col_for_pattern(this_pattern)
          redonly      = this_col_pat[2, ]=='red'
          if(sum(redonly) > 0)
          {
            text(x = contmarker[1:tlen, 3][redonly], y = rep(ipat*floor(totalpollennum/npollen), sum(redonly)), labels = this_seq[redonly], col = this_col_pat[2, ][redonly] , family = "Helvetica", cex=labelsizecex) 
            #points(x = contmarker[1:tlen, 3][redonly], y = rep(ipat*floor(totalpollennum/npollen), sum(redonly)), type = 'p', col = this_col_pat[2, ][redonly], pch=19, cex=labelsizecex)
          }
          text(x = xlimend+round(xlimend*0.1), y = rep(ipat*floor(totalpollennum/npollen), 1), labels = mcluster_pattern[ipat, 2], col = "red", family = "Helvetica", cex=labelsizecex)          
        }    
        # then blue: wrong genotype from paternal
        for(ipat in 1:npollen)
        {
          this_seq     = strsplit( tolower( as.vector(mcluster_seq[ipat, 1]) ), "")[[1]]
          this_pattern = as.vector(mcluster_pattern[ipat, 1])
          tlen         = nchar(this_pattern)
          this_col_pat = get_col_for_pattern(this_pattern)
          darkblueonly = this_col_pat[2, ]=='darkblue'
          if(sum(darkblueonly) > 0)
          {
            text(x = contmarker[1:tlen, 3][darkblueonly], y = rep(ipat*floor(totalpollennum/npollen), sum(darkblueonly)), labels = this_seq[darkblueonly], col = "darkblue", family = "Helvetica", cex=labelsizecex+0.1) 
            #points(x = contmarker[1:tlen, 3][darkblueonly], y = rep(ipat*floor(totalpollennum/npollen), sum(darkblueonly)), type = 'p', col = "darkblue", pch=4, cex=labelsizecex+0.1)        
          }
        }        
      }  
    }
    ############################################## visualize pollen patterns in maternal cluster
    plot(1, 1, xlim = c(xlimsta, xlimend+round(xlimend*0.2) ), ylim = c(0, totalpollennum), frame.plot = F, axes = F, xlab = "", ylab = "", col="white")
    segments(x0 = contmarker$V3, y0 = 0, x1 = contmarker$V3, y1 = totalpollennum, col="gray", lwd=0.3)
    title(main="Pollens classified in paternal cluster", cex.main = 1, col.main="black")
    if(is.null(pcluster_pattern) == FALSE & pcluster_pattern!=0)
    {
      npollen = length(pcluster_pattern$V1)
      if(npollen > 0)
      {
        # first darkblue: paternal truth
        for(ipat in 1:npollen)
        {
          this_seq     = strsplit( tolower(as.vector(pcluster_seq[ipat, 1]) ), "")[[1]]
          this_pattern = as.vector(pcluster_pattern[ipat, 1])
          tlen         = nchar(this_pattern)
          this_col_pat = get_col_for_pattern(this_pattern)
          darkblueonly = this_col_pat[2, ]=='darkblue'
          if(sum(darkblueonly) > 0)
          {
            text(x = contmarker[1:tlen, 3][darkblueonly], y = rep(ipat*floor(totalpollennum/npollen), sum(darkblueonly)), labels = this_seq[darkblueonly], col = this_col_pat[2, ][darkblueonly] , family = "Helvetica", cex=labelsizecex) 
            #points(x = contmarker[1:tlen, 3][darkblueonly], y = rep(ipat*floor(totalpollennum/npollen), sum(darkblueonly)), type = 'p', col = this_col_pat[2, ][darkblueonly], pch=19, cex=labelsizecex)
          }
          text(x = xlimend+round(xlimend*0.1), y = rep(ipat*floor(totalpollennum/npollen), 1), labels = pcluster_pattern[ipat, 2], col = "darkblue", family = "Helvetica", cex=labelsizecex)          
        }  
        # then red: wrong genotype from maternal
        for(ipat in 1:npollen)
        {
          this_seq     = strsplit( tolower(as.vector(pcluster_seq[ipat, 1]) ), "")[[1]]
          this_pattern = as.vector(pcluster_pattern[ipat, 1])
          tlen         = nchar(this_pattern)
          this_col_pat = get_col_for_pattern(this_pattern)
          redonly = this_col_pat[2, ]=='red'
          if(sum(redonly) > 0)
          {
            text(x = contmarker[1:tlen, 3][redonly], y = rep(ipat*floor(totalpollennum/npollen), sum(redonly)), labels = this_seq[redonly], col = "red", family = "Helvetica", cex=labelsizecex+0.1) 
            #points(x = contmarker[1:tlen, 3][redonly], y = rep(ipat*floor(totalpollennum/npollen), sum(redonly)), type = 'p', col = "red", pch=4, cex=labelsizecex+0.1) 
          }
        }        
      }
    }
    ############################################## visualize pollen patterns in unknown cluster
    bfname <- basename(clusterfile)
    plot(1, 1, xlim = c(xlimsta, xlimend+round(xlimend*0.2) ), ylim = c(0, totalpollennum), frame.plot = F, axes = F, xlab = "", ylab = "", col="white")
    segments(x0 = contmarker$V3, y0 = 0, x1 = contmarker$V3, y1 = totalpollennum, col="gray", lwd=0.3)
    if(length(this_size)<0) this_size = "Not Found!"
    title(main=paste("Pollens un-classified: ", bfname, "\nsize: ", this_size, " bp", sep=""), cex.main = 1, col.main="gray")
    if(is.null(ucluster_pattern) == FALSE & ucluster_pattern!=0)
    {
      npollen = length(ucluster_pattern$V1)
      if(npollen > 0)
      {
        for(ipat in 1:npollen)
        {
          this_seq     = strsplit( tolower(as.vector(ucluster_seq[ipat, 1]) ), "")[[1]]
          this_pattern = as.vector(ucluster_pattern[ipat, 1])
          tlen         = nchar(this_pattern)
          this_col_pat = get_col_for_pattern(this_pattern)
          nongrayonly  = this_col_pat[2, ]!='gray'
          if(sum(nongrayonly) > 0)
          {
            text(x = contmarker[1:tlen, 3][nongrayonly], y = rep(ipat*floor(totalpollennum/npollen), sum(nongrayonly) ), labels = this_seq[nongrayonly], col = this_col_pat[2, ][nongrayonly], family = "Helvetica", cex=labelsizecex)  
            #points(x = contmarker[1:tlen, 3][nongrayonly], y = rep(ipat*floor(totalpollennum/npollen), sum(nongrayonly)), type = 'p', col = this_col_pat[2, ][nongrayonly], pch=19, cex=labelsizecex)
          }
          text(x = xlimend+round(xlimend*0.1), y = rep(ipat*floor(totalpollennum/npollen), 1), labels = ucluster_pattern[ipat, 2], col = "gray", family = "Helvetica", cex=labelsizecex)          
        }
      }
    }
    ###
    dev.off()
  }
}